pkg/abstractions/shell/shell.go

```diff
@@ -0,0 +1,253 @@
+package shell
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/labstack/echo/v4"
+
+	abstractions "github.com/beam-cloud/beta9/pkg/abstractions/common"
+	pb "github.com/beam-cloud/beta9/proto"
+
+	"github.com/beam-cloud/beta9/pkg/auth"
+	"github.com/beam-cloud/beta9/pkg/common"
+	"github.com/beam-cloud/beta9/pkg/network"
+	"github.com/beam-cloud/beta9/pkg/repository"
+	"github.com/beam-cloud/beta9/pkg/scheduler"
+	"github.com/beam-cloud/beta9/pkg/types"
+)
+
+const (
+	shellRoutePrefix       string = "/shell"
+	shellContainerPrefix   string = "shell"
+	defaultContainerCpu    int64  = 100
+	defaultContainerMemory int64  = 128
+)
+
+type ShellService interface {
+	pb.ShellServiceServer
+	CreateShell(ctx context.Context, in *pb.CreateShellRequest) (*pb.CreateShellResponse, error)
+}
+
+type SSHShellService struct {
+	pb.UnimplementedShellServiceServer
+	ctx             context.Context
+	config          types.AppConfig
+	rdb             *common.RedisClient
+	keyEventManager *common.KeyEventManager
+	scheduler       *scheduler.Scheduler
+	backendRepo     repository.BackendRepository
+	workspaceRepo   repository.WorkspaceRepository
+	containerRepo   repository.ContainerRepository
+	eventRepo       repository.EventRepository
+	tailscale       *network.Tailscale
+}
+
+type ShellServiceOpts struct {
+	Config        types.AppConfig
+	RedisClient   *common.RedisClient
+	BackendRepo   repository.BackendRepository
+	WorkspaceRepo repository.WorkspaceRepository
+	ContainerRepo repository.ContainerRepository
+	Scheduler     *scheduler.Scheduler
+	RouteGroup    *echo.Group
+	Tailscale     *network.Tailscale
+	EventRepo     repository.EventRepository
+}
+
+func NewSSHShellService(
+	ctx context.Context,
+	opts ShellServiceOpts,
+) (ShellService, error) {
+	keyEventManager, err := common.NewKeyEventManager(opts.RedisClient)
```

Comment:
TODO: use key event manage to detect existing (running) shell containers on start